nvdaHelper/localWin10/oneCoreSpeech.cpp

```diff
@@ -30,6 +37,130 @@ using namespace winrt::Windows::Media;
 
+std::recursive_mutex InstanceManager::_instanceStateMutex;
+std::vector<OcSpeech*> InstanceManager::_terminatedInstances;
+std::unique_ptr<OcSpeech> InstanceManager::_instance;
+std::atomic<InstanceState> InstanceManager::_instanceState = InstanceState::notInitialized;
+std::condition_variable_any InstanceManager::_readyForInitialization;
+std::atomic_int SpeakThreadGuard::_speechThreads = 0;
+
+SpeakThreadGuard::SpeakThreadGuard() {
+	/*
+	When initialized, increases the count of active speech threads
+	waiting on a callback.
+	When exiting the scope of initialization, decreases the count
+	of active speech threads and deletes the OneCore instance if it has
+	been terminated and no speech callbacks are waiting.
+	*/
+	++_speechThreads;
+}
+
+bool SpeakThreadGuard::areCallbacksPending() {
+	/*
+	Checks if any speak threads are active, i.e. with callbacks pending.
+	*/
+	return _speechThreads != 0;
+}
+
+SpeakThreadGuard::~SpeakThreadGuard() {
+	/* 
+	When exiting the scope of initialization, decreases the count
+	of active speech threads and deletes the OneCore instance if it has
+	been terminated and no speech callbacks are waiting.
+	*/
+	--_speechThreads;
+	if (_speechThreads < 0) {
+		throw runtime_error(
+			"Negative callback count."
+			"decreasePendingCount has been called too many times."
+		);
+	}
+	InstanceManager::deleteInstanceIfTerminatedAndReady();
+}
+
+void InstanceManager::deleteInstanceIfTerminatedAndReady() {
+	/*
+	If the instance is terminated and no speak callbacks are pending,
+	delete the instance and notify any waiting threads,
+	such as the instance initializer.
+	*/
+	std::lock_guard g(_instanceStateMutex);
+	if (
+		_instanceState == InstanceState::terminated
+		&& !SpeakThreadGuard::areCallbacksPending()
+	) {
+		_instance.reset();
+		_instanceState = InstanceState::notInitialized;
+		_readyForInitialization.notify_all();
+	}
+}
+
+void InstanceManager::waitUntilReadyForInitialization() {
+	/* Wait for a signal that there is no active or terminated instance. */
+	std::unique_lock lock(_instanceStateMutex);
+	_readyForInitialization.wait(lock, []{
+		return !_instance && _instanceState == InstanceState::notInitialized;
+	});
+}
+
+OcSpeech* InstanceManager::getActiveInstance(OcSpeech* token) {
+	/* Throw a runtime error if no instance is active */
+	_assertInstanceActive(token);
+	return _instance.get();
+}
+
+void InstanceManager::_assertInstanceActive(OcSpeech* token) {
+	if (_instanceState != InstanceState::active) {
+		throw runtime_error("Supplied OneCore token is not active");
+	}
+	if (_instance.get() != token) {
+		throw runtime_error("Supplied OneCore instance token does not match initialized instance");
+	}
+}
+
+OcSpeech* InstanceManager::initializeNewInstance() {
+	/*
+	Initializes a new OneCore instance.
+	Waits until an instance has been fully terminated,
+	and then (once speech callbacks have finished) deleted.
+	*/
+	std::lock_guard g(_instanceStateMutex);

```

Comment:
This can cause a deadlock. This is acquiring the lock potentially while there is an active `SpeakThreadGuard` due to `OcSpeech::speak` which can't terminate until it acquires the lock to delete the instance and signal `_readyForInitialization`

Instead, I'd make `waitUntilReadyForInitialization` transfer ownership of the lock back to the caller after it is acquired safely.